web/crux-ui/src/server/registry-api/google-api-client.ts

```diff
@@ -0,0 +1,87 @@
+import { RegistryImageTags } from '@app/models'
+import { internalError, unauthorizedError } from '@server/error-middleware'
+import { JWT } from 'google-auth-library'
+import { GetAccessTokenResponse } from 'google-auth-library/build/src/auth/oauth2client'
+import { RegistryApiClient } from './registry-api-client'
+
+export type GoogleClientOptions = {
+  username?: string
+  password?: string
+}
+
+export class GoogleRegistryClient implements RegistryApiClient {
+  private headers: HeadersInit = {}
+  private client: JWT
+
+  constructor(private url: string, options?: GoogleClientOptions) {
+    if (options?.username) {
+      if (!options.password) {
+        throw unauthorizedError(`Invalid authentication parameters for: ${url}`)
+      }
+
+      this.client = new JWT({
+        email: options.username,
+        key: options.password,
+        scopes: ['https://www.googleapis.com/auth/cloud-platform'],
+      })
+    }
+  }
+
+  async registryCredentialsToBearerAuth() {
+    let accessToken: GetAccessTokenResponse
+    try {
+      accessToken = await this.client.getAccessToken()
+    } catch (e) {
+      throw unauthorizedError(`Google auth request failed`)
+    }
+
+    this.headers = {
+      Authorization: `Bearer ${accessToken?.token}`,
+      Accept: `application/json`,
+    }
+  }
+
+  async catalog(text: string, take: number): Promise<string[]> {
+    if (this.client) {
+      await this.registryCredentialsToBearerAuth()
+    }
+
+    const url = `https://gcr.io/v2/${this.url}/tags/list`
+
+    const res = await fetch(url, {
+      headers: this.headers,
+    })
+
+    if (!res.ok) {
+      const errorMessage = `Google repositories request failed with status: ${res.status} ${res.statusText}`
+      throw res.status === 401 ? unauthorizedError(errorMessage) : internalError(errorMessage)
+    }
+
+    const json = (await res.json()) as { child: string[] }
+
+    return json.child.filter(it => it.includes(text)).slice(0, take)
+  }
+
+  async tags(image: string): Promise<RegistryImageTags> {
+    if (this.client) {
+      await this.registryCredentialsToBearerAuth()
+    }
+
+    const url = `https://gcr.io/v2/${this.url}/${image}/tags/list`
+
+    const tagRes = await fetch(url, {
+      headers: this.headers,
+    })
+
+    if (!tagRes.ok) {
+      const errorMessage = `Google tags request failed with status: ${tagRes.status} ${tagRes.statusText}`
+      throw tagRes.status === 401 ? unauthorizedError(errorMessage) : internalError(errorMessage)
```

Comment:
Are you sure in every scenario if not 401, we always got `internalError()`?

Reply:
At this point, our request failed and there will be only one informative error that we can handle (unauthorizedError), everything else will be uninformative that's why I wrapped them into the internalError

Reply:
Okey I understand it, but If we got for example 404 you throw internal error which isn't valid. Any idea to improv this code @m8vago ? What do you think?

Reply:
RegistryAccessValidationGuard will test the registry at saving, for example it will prevent the save when the status code is 404.